feat: implement timezone functionality

Ethan225300 opened the pull request:
- Created the Timezone model: Defined the structure for storing timezone data in the database.
- Implemented TimezoneService: Developed methods to retrieve all available timezones and update user-specific timezone settings.
- Added API endpoints for fetching all timezones and updating the user's timezone.
- Handled GraphQL queries and mutations related to timezone data.
- Added relative tests 
- Resolve CP-31
<img width="1254" alt="451f0d1947a592a8b9dd2738a574f2f" src="https://github.com/BeeQuantAI/platform_api/assets/162969330/bdcf4d6f-6fda-4845-930e-75963d564b08">
<img width="1244" alt="9cf795ee4f361d4aa86039cedbc2382" src="https://github.com/BeeQuantAI/platform_api/assets/162969330/f6bf1256-067a-4983-a35a-e29cd5f25f8c">
<img width="1237" alt="09508e9b739984d4c97c7b4f4d03100" src="https://github.com/BeeQuantAI/platform_api/assets/162969330/b6971546-a0aa-4c33-b783-d91ef8a80cc6">

Chris-Wang commented on src/app.controller.spec.ts:
file gets any formatting issue resulting every line gets updated?

Please fix this issue, it should only show updated/changed lines

Chris-Wang commented on src/app.controller.ts:
file gets any formatting issue resulting every line gets updated?

Please fix this issue, it should only show updated/changed lines

Chris-Wang commented on src/modules/user/models/user.entity.ts:
file gets any formatting issue resulting every line gets updated?

Please fix this issue, it should only show updated/changed lines

Chris-Wang commented on src/modules/user/user.resolver.ts:
file gets any formatting issue resulting every line gets updated?

Please fix this issue, it should only show updated/changed lines

Chris-Wang commented on src/modules/user/user.resolver.ts:
why update `lastLoginTime` logic spread all over most of functions?

shouldn't this filed only be updated after user login? as opposite to someone did a query of that user

Chris-Wang commented on src/modules/user/user.service.ts:
file gets any formatting issue resulting every line gets updated?

Please fix this issue, it should only show updated/changed lines

Chris-Wang commented on src/modules/user/user.service.ts:
again, why update lastLoginTime logic spread all over most of functions?

Chris-Wang commented on src/timezone/timezone.controller.ts:
we don't use controller as our graphql services

Chris-Wang commented on src/timezone/timezone.service.ts:
```diff
@@ -0,0 +1,36 @@
+  constructor(private readonly userService: UserService) {}
+
+  getAllTimezones(): ITimezone[] {
+    return moment.tz.names().map(name => ({
```
do we really need to return all timezones?

Chris-Wang reviewed (changes requested):
We don't use controllers for our graphql services, please reimplemented the timezone features

Also, please attach updated gql queries and testing result on playground

Chris-Wang commented on src/modules/user/user.service.ts:
```diff
@@ -42,27 +43,39 @@ export class UserService {
-    return res;
+    const users = await this.userRepository.find();
+    users.forEach(user => {
+      if (user.lastLoginTime && user.timezone) {
```
the last login time and timezone won't be changed after user queries being executed

Chris-Wang commented on src/modules/user/user.service.ts:
```diff
@@ -42,27 +43,39 @@ export class UserService {
-    });
-    return res;
+    const user = await this.userRepository.findOne({ where: { id } });
+    if (user && user.lastLoginTime && user.timezone) {
```
the last login time and timezone won't be changed after user queries being executed

Chris-Wang commented on src/modules/user/user.service.ts:
```diff
@@ -42,27 +43,39 @@ export class UserService {
-      },
-    });
-    return res;
+    const user = await this.userRepository.findOne({ where: { email } });
```
the last login time and timezone won't be changed after user queries being executed

Chris-Wang commented on src/timezone/timezone.service.ts:
```diff
@@ -0,0 +1,48 @@
+    return moment.tz.names().includes(timezone);
+  }
+
+  async updateUserTimezone(userId: string, timezone: string): Promise<any> {
```
consider update user is a private service, it needs to validate user's authentication, then allow user to update timezone

Chris-Wang reviewed (changes requested):
remember to resolve comments

Chris-Wang reviewed (changes requested):
- Consider using GqlAuthGuard, we don't need timezone service to validateUserToken

hint: check how we did on updating user profile

Ethanwyf1 committed:
feat: implement timezone functionality
- Created the Timezone model: Defined the structure for storing timezone data in the database.
- Implemented TimezoneService: Developed methods to retrieve all available timezones and update user-specific timezone settings.
- Added API endpoints for fetching all timezones and updating the user's timezone.
- Handled GraphQL queries and mutations related to timezone data.
- Added relative tests
Resolve CP-31
